### PR TITLE
Dynamic programming: Can_sum

### LINKED ISSUE #28: Challenge: can_sum

### PR DESCRIPTION
**Pull request by delvor**
- [x] PR's title includes the challenge's name and language
- [x] PR has short and clear description of the challenge
- [x] PR has appropriate labels and milestones for easy identification
- [ ] PR it is assigned to the owner
- [x] reviewers are assigned
- [ ] the PR contributes only one focused change
- [x] the branch is up to date with `main`/`master`
- [x] the code works when pulled and run locally
- [ ] all conflicts are resolved (if any)
- [ ] It is linked to an issue in the appropriate column of the project board
- [ ] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [x] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [ ] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [x] There is a module header
- [x] There is a module docstring
- [x] File names are in snake_case
- [x] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [x] Parameter description
- [x] Return value description
- [ ] Include any assumptions
- [ ] Include 3 or more (passing!) doctests
- [ ] Include 1-2 use cases (if necessary)

## Function Implementation

- [x] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [x] The code is auto-formatted
- [ ] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [x] The code follows the strategy as simply as possible
- [ ] Variable names are clear and helpful
- [ ] Comments explain the strategy (if necessary)
- [x] There are type annotations
- [ ] (challenge) The code includes defensive assertions

## Unit Test Suite

- [x] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Each unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases
- [ ] Test suite includes black-box and glass-box tests

### CODE REVIEW

File: src/dynamic_programming/python/can_sum/src/can_sum.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    True """
+    if number < 0:
+        return False
+    elif number == 0:
+        return True
+
+    for i in arr:
+        reminder = number-i
+
+        return can_sum_basic(reminder, arr[1:]) or can_sum_basic(number, arr[1:])
```
Comment (tvsirius) [inline src/dynamic_programming/python/can_sum/src/can_sum.py:31]: should not we pass to recursion array without i element
Just don't get why you are passing arr[1:]

you could take some element twise with this approach

File: src/dynamic_programming/python/can_sum/src/can_sum.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    True """
+
+    if __memo > number:
+        return False
+    elif number == __memo:
+        return True
+    for element in arr:
+        __memo += element
+        return can_sum_memo(number, arr[1:],  __memo) or can_sum_memo(number, arr[1:])
+
```
Comment (tvsirius) [inline src/dynamic_programming/python/can_sum/src/can_sum.py:55]: the same

File: src/dynamic_programming/python/can_sum/src/can_sum.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    if number < 0:
+        return False
+    elif number == 0:
+        return True
+
+    for i in arr:
+        reminder = number-i
+
+        return can_sum_basic(reminder, arr[1:]) or can_sum_basic(number, arr[1:])
+    return False
```
Comment (tvsirius) [inline src/dynamic_programming/python/can_sum/src/can_sum.py:32]: I see now

    for i in arr:
   
does not make any sence - it's confusing

as this last return False
because you use
        return can_sum_basic(reminder, arr[1:]) or can_sum_basic(number, arr[1:])
function will return and nor 'for' loop, nor last ruturn will excecute

so you could write:
```
i=arr[0]
reminder = number-i
return can_sum_basic(reminder, arr[1:]) or can_sum_basic(number, arr[1:])
```

this will make code much better

File: src/dynamic_programming/python/can_sum/src/can_sum.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    Parameters:
+        arr: list[int] - List of items to sum
+        number:int - Target sum
+        memo:int - memorization for recurcive call only
+
+    Returns -> bool    
+
+    >>> can_sum_memo([3,1,1],2)
+    True """
+
```
Comment (tvsirius) [inline src/dynamic_programming/python/can_sum/src/can_sum.py:47]: Maybe comments will help to understang how this recursion is working

**Review by tvsirius — CHANGES_REQUESTED** [review]
Good code, works, but please fix the issue

No assertions also
Not much comments on logic

File: src/dynamic_programming/python/can_sum/src/can_sum.py (hunk 1 above)
Comment (delvor) [reply src/dynamic_programming/python/can_sum/src/can_sum.py:31]: because we need to slice array to check variants

File: src/dynamic_programming/python/can_sum/src/can_sum.py (hunk 3 above)
Comment (delvor) [reply src/dynamic_programming/python/can_sum/src/can_sum.py:32]: this not going to work. second iteration will be next number  of array and you will check everything to [0] index of array

File: src/dynamic_programming/python/can_sum/src/can_sum.py (hunk 4 above)
Comment (delvor) [reply src/dynamic_programming/python/can_sum/src/can_sum.py:47]: pretty muck same as previous but use memo

File: src/dynamic_programming/python/can_sum/src/can_sum.py (hunk 3 above)
Comment (tvsirius) [reply src/dynamic_programming/python/can_sum/src/can_sum.py:32]: Yes, I forgot the case when len(arr)==0, then iteration will not start, and you will gor to line 32 and return False, it a zero case

But the for here is not needed. So it's no problem to merge, but it's kind of mistake and code confusion

```
for i in range(1,10):
    j=some_function(i)*100
    return j
```
is totally equvalent to:
```
i=1
j=some_function(i)*100
return j
````

after retun loop will not run

I rewrote your code this way and it works:
```
    # for i in arr:
    #     reminder = number-i

    #     return can_sum_basic(reminder, arr[1:]) or can_sum_basic(number, arr[1:])
    
    # Zero case
    if len(arr)==0:
        return False

    return can_sum_basic(number-arr[0], arr[1:]) or can_sum_basic(number, arr[1:])
    # return False

```

File: src/dynamic_programming/python/can_sum/src/can_sum.py (hunk 4 above)
Comment (tvsirius) [reply src/dynamic_programming/python/can_sum/src/can_sum.py:47]: the pervios (without the memo) also lack comments

File: src/dynamic_programming/python/can_sum/src/can_sum.py (hunk 3 above)
Comment (delvor) [reply src/dynamic_programming/python/can_sum/src/can_sum.py:32]: your solution is your solution. im not going to play wit it more. font like it - you can write your own solution.

**Comment by delvor** [comment]
open any lib code. there is not much comment on logic if it not something special. no need to be coment nut

File: src/dynamic_programming/python/can_sum/src/can_sum.py (hunk 3 above)
Comment (tvsirius) [reply src/dynamic_programming/python/can_sum/src/can_sum.py:32]: This is not about the solution or whose code is better, but as you wish

**Comment by delvor** [comment]
Ok. Do it yourself
